process/coordinator/process.go

```diff
@@ -1162,6 +1179,147 @@ func getNumOfCrossShardScCallsOrSpecialTxs(
+) error {
+	var err error
+	var gasConsumedByTxInReceiverShard uint64
+	gasConsumedByMiniBlockInReceiverShard := uint64(0)
+
+	for _, txHash := range miniBlock.TxHashes {
+		txHandler, ok := mapHashTx[string(txHash)]
+		if !ok {
+			log.Debug("missing transaction in checkGasConsumedByMiniBlockInReceiverShard ", "type", miniBlock.Type, "txHash", txHash)
+			return process.ErrMissingTransaction
+		}
+
+		_, txTypeDstShard := tc.txTypeHandler.ComputeTransactionType(txHandler)
+		if txTypeDstShard == process.MoveBalance {
+			gasConsumedByTxInReceiverShard = tc.economicsFee.ComputeGasLimit(txHandler)
+		} else {
+			gasConsumedByTxInReceiverShard = txHandler.GetGasLimit()
+		}
+
+		gasConsumedByMiniBlockInReceiverShard, err = core.SafeAddUint64(gasConsumedByMiniBlockInReceiverShard, gasConsumedByTxInReceiverShard)
+		if err != nil {
+			return err
+		}
+	}
+
+	if gasConsumedByMiniBlockInReceiverShard > tc.economicsFee.MaxGasLimitPerBlock(miniBlock.ReceiverShardID) {
+		return process.ErrMaxGasLimitPerMiniBlockInReceiverShardIsReached
+	}
+
+	return nil
+}
+
+func (tc *transactionCoordinator) verifyFees(
+	header data.HeaderHandler,
+	body *block.Body,
+	mapMiniBlockTypeAllTxs map[block.Type]map[string]data.TransactionHandler,
+) error {
+	totalMaxAccumulatedFees := big.NewInt(0)
+	totalMaxDeveloperFees := big.NewInt(0)
+
+	for _, miniBlock := range body.MiniBlocks {
+		if miniBlock.Type == block.PeerBlock {
+			continue
+		}
+
+		maxAccumulatedFeesFromMiniBlock, maxDeveloperFeesFromMiniBlock, err := tc.getMaxAccumulatedAndDeveloperFees(
+			miniBlock,
+			mapMiniBlockTypeAllTxs[miniBlock.Type],
+		)
+		if err != nil {
+			return err
+		}
+
+		totalMaxAccumulatedFees.Add(totalMaxAccumulatedFees, maxAccumulatedFeesFromMiniBlock)
+		totalMaxDeveloperFees.Add(totalMaxDeveloperFees, maxDeveloperFeesFromMiniBlock)
+	}
+
+	if header.GetAccumulatedFees().Cmp(totalMaxAccumulatedFees) > 0 {
+		return process.ErrMaxAccumulatedFeesExceeded
+	}
+	if header.GetDeveloperFees().Cmp(totalMaxDeveloperFees) > 0 {
+		return process.ErrMaxDeveloperFeesExceeded
+	}
+
+	return nil
+}
+
+func (tc *transactionCoordinator) getMaxAccumulatedAndDeveloperFees(
+	miniBlock *block.MiniBlock,
+	mapHashTx map[string]data.TransactionHandler,
+) (*big.Int, *big.Int, error) {
+	maxAccumulatedFeesFromMiniBlock := big.NewInt(0)
+	maxDeveloperFeesFromMiniBlock := big.NewInt(0)
+
+	for _, txHash := range miniBlock.TxHashes {
+		txHandler, ok := mapHashTx[string(txHash)]
+		if !ok {
+			log.Debug("missing transaction in getMaxAccumulatedFeesAndDeveloperFees ", "type", miniBlock.Type, "txHash", txHash)
+			return big.NewInt(0), big.NewInt(0), process.ErrMissingTransaction
+		}
+
+		maxAccumulatedFeesFromTx := core.SafeMul(txHandler.GetGasLimit(), txHandler.GetGasPrice())
+		maxAccumulatedFeesFromMiniBlock.Add(maxAccumulatedFeesFromMiniBlock, maxAccumulatedFeesFromTx)
+
+		maxDeveloperFeesFromTx := core.GetPercentageOfValue(maxAccumulatedFeesFromTx, tc.economicsFee.DeveloperPercentage())
```

Comment:
Heads up, this function is changed in https://github.com/ElrondNetwork/elrond-go/pull/2760

Reply:
I will check it after #2760 merge :+1: